core/src/main/java/io/grpc/internal/PickFirstLeafLoadBalancer.java

```diff
@@ -0,0 +1,492 @@
+
+      // start connection attempt at first address but stay in transient failure
+      if (currentState == TRANSIENT_FAILURE) {
+        requestConnection();
+      // start connection attempt at first address when requested
+      } else if (currentState == IDLE) {
+        SubchannelPicker picker =
+            new RequestConnectionPicker(this);
+        updateBalancingState(IDLE, picker);
+      // start connection attempt at first address
+      } else if (currentState == CONNECTING
+          || (currentState == READY && !addressIndex.seekTo(previousAddress))) {
+        addressIndex.reset();
+        updateBalancingState(CONNECTING, new Picker(PickResult.withNoResult()));
+        requestConnection();
+      }
+    }
+    return true;
+  }
+
+  @Override
+  public void handleNameResolutionError(Status error) {
+    for (Subchannel subchannel : subchannels.values()) {
+      subchannel.shutdown();
+    }
+    subchannels.clear();
+    // NB(lukaszx0) Whether we should propagate the error unconditionally is arguable. It's fine
+    // for time being.
+    updateBalancingState(TRANSIENT_FAILURE, new Picker(PickResult.withError(error)));
+  }
+
+  void processSubchannelState(Subchannel subchannel, ConnectivityStateInfo stateInfo) {
+    ConnectivityState newState = stateInfo.getState();
+    // If we have shutdown channels or updated addresses, we may still receive state updates
+    // from subchannels. To prevent pickers from returning these obselete subchannels, this logic
+    // is included to check if the current list of active subchannels includes the current one.
+    if (!subchannels.containsValue(subchannel)) {
+      return;
+    }
+    if (newState == SHUTDOWN) {
+      return;
+    }
+    if (newState == IDLE) {
+      helper.refreshNameResolution();
+    }
+    // If we are transitioning from a TRANSIENT_FAILURE to CONNECTING or IDLE we ignore this state
+    // transition and still keep the LB in TRANSIENT_FAILURE state. This is referred to as "sticky
+    // transient failure". Only a subchannel state change to READY will get the LB out of
+    // TRANSIENT_FAILURE. If the state is IDLE we additionally request a new connection so that we
+    // keep retrying for a connection.
+
+    // With the new pick first implementation, iterative requests for connections will not be
+    // requested once the first pass through is complete. This means that individual subchannels
+    // are responsible for coming out of backoff and starting a transport.
+    // However, if the first pass through is complete and we report TRANSIENT_FAILURE but an address
+    // update occurs, the iterative logic will still be present for the first pass through for
+    // the new address list even though we are in a state of TRANSIENT_FAILURE.
+    if (currentState == TRANSIENT_FAILURE) {
+      if (!iterate && (newState == CONNECTING || newState == TRANSIENT_FAILURE)) {
+        // each subchannel is responsible for its own backoff
+        return;
```

Comment:
From meeting: Include state transition from TF --> IDLE, request a connection at that point. We likely won't encounter this case much, if at all.